src/python/pants/goal/run_tracker.py

```diff
@@ -429,46 +429,44 @@ def create_dict_with_nested_keys_and_val(all_keys, value, index):
     elif index == 0:
       return {all_keys[index]: value}
 
-
   @staticmethod
-  def merge_list_of_keys_into_dict(data, keys, value, index):
-    """ Recursively merge list of keys that points to the given value into data.
+  def merge_list_of_keys_into_dict(data, keys, value, index=0):
+    """Recursively merge list of keys that points to the given value into data.
+
+    Will not overwrite data that is already associated with a given key list.
 
     :param dict data: Dictionary to be updated.
     :param list of string keys: The keys that point to where the value should be stored.
            Will recursively find the correct place to store in the nested dicts.
     :param primitive value: The value of the information being stored.
-    :param int index: The index into the list of keys (starting form the beginning). This should
-           always be 0 when being called outside of this function.
+    :param int index: The index into the list of keys (starting form the beginning).
     """
     if len(keys) == 0:
       raise ValueError('Keys must contain at least one key')
     if len(keys) < 2 or not data:
       new_data_to_add = RunTracker.create_dict_with_nested_keys_and_val(keys, value, len(keys) - 1)
       data.update(new_data_to_add)
       return
+
+    this_keys_contents = data.get(keys[index])
+    if this_keys_contents:
+      if isinstance(this_keys_contents, dict):
+        RunTracker.merge_list_of_keys_into_dict(this_keys_contents, keys, value, index + 1)
     else:
-      this_keys_contents = data.get(keys[index])
-      if this_keys_contents:
-        if isinstance(this_keys_contents, dict):
-          new_keys = keys[index+1:]
-          RunTracker.merge_list_of_keys_into_dict(this_keys_contents, new_keys, value, 0)
-        else:
-          new_keys = keys[index+1:]
-          new_data_to_add = RunTracker.create_dict_with_nested_keys_and_val(new_keys, value, len(new_keys) - 1)
-          data[keys[index]] = new_data_to_add
-          return
-      else:
-        new_keys = keys[index:]
-        new_data_to_add = RunTracker.create_dict_with_nested_keys_and_val(new_keys, value, len(new_keys) - 1)
-        data.update(new_data_to_add)
-        return
+      new_keys = keys[index:]
+      new_data_to_add = RunTracker.create_dict_with_nested_keys_and_val(new_keys, value,
+        len(new_keys) - 1)
+      data.update(new_data_to_add)
+      return
```

Comment:
Redundant statement... since the function doesn't return anything, no need to explicitly return.